http-webhook.md

```diff
@@ -0,0 +1,339 @@
+and a registration handshake that protects the sender from being abused for
+flooding arbitrary HTTP sites with requests.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][RFC2119].
+
+### 1.2. Relation to HTTP
+
+This specification prescribes rules constraining the use and handling of
+specific [HTTP methods][RFC7231-Section-4] and headers.
+
+This specification also applies equivalently to HTTP/2 ([RFC7540][RFC7540]),
+which is compatible with HTTP 1.1 semantics.
+
+## 2. Delivering notifications
+
+### 2.1. Delivery request
+
+Notifications are delivered using a HTTP request. The response indicates the
+resulting status of the delivery.
+
+HTTP-over-TLS (HTTPS) [RFC2818][RFC2818] MUST be used for the connection.
+
+The HTTP method for the delivery request MUST be [POST][POST].
+
+The [`Content-Type`][Content-Type] header MUST be carried and the request MUST
+carry a notification payload of the given content type. Requests without
+payloads, e.g. where the notification is entirely expressed in HTTP headers,
+are not permitted by this specification.
+
+This specification does not further constrain the content of the notification,
+and it also does not prescribe the [HTTP target resource][RFC7230-Section-5-1]
+that is used for delivery.
+
+If the delivery target supports and requires [Abuse Protection][#4-abuse-protection],
+the delivery request MUST include the `Origin` header. The `Origin` header value
+is a DNS name expression that identifies the sending system.
+
+### 2.2. Delivery response
+
+The delivery response MAY contain a payload providing detail status
+information in the case of handling errors. This specification does not attempt
+to define such a payload.
+
+The response MUST NOT use any of the [3xx HTTP Redirect status codes][3xx] and
+the client MUST NOT follow any such redirection.
+
+If a delivery target has been retired, but the HTTP site still exists, the
+site SHOULD return a [410 "Gone"][410] status code and the sender SHOULD refrain
+from sending any further notifications.
+
+If the delivery target is unable to process the request due to exceeding a
+request rate limit, it SHOULD return a [429 Too Many Requests][429] status code
+and MUST include the [`Retry-After`][Retry-After] header. The sender MUST
+observe the value of the Retry-After header and refrain from sending further
+requests until the indicated time.
+
+If the delivery has been accepted and processed, and if the response carries a
+payload with processing details, the response MUST have the [200 OK][200] or
+[201 Created][201] status code. In this case, the response MUST carry a
+[`Content-Type`][Content-Type] header.
+
+If the delivery has been accepted and processed, but carries no payload, the
+response MUST have the [201 Created][201] or [204 No Content][204] status code.
```

Comment:
A number of new web developers send 200 with Content-Length 0. Is there a reason to explicitly call these non-compliant?

Reply:
Because they are not compliant. HTTP reserves 204 for when you have no payload. I can't relax the rules of HTTP here.

Reply:
@clemensv according to the HTTP spec any Content-Length greater than or equal to 0 is a valid value. See [here](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.13)

Reply:
@inlined @glennblock the normative spec for that is now RFC7230 which has superseded RFC2616. The [relevant section is here](https://tools.ietf.org/html/rfc7230#section-3.3.2). I agree that the wording distinguishes between an absent and an empty body in the context of allowing for POST with no payload. There is also this SHOULD rule:

"A user agent SHOULD NOT send a Content-Length header field when the request message does not contain a payload body and the method semantics do not anticipate such a body."

The HTTP spec isn't clear here about the recommendation for responses where there is no payload. 

My position is that Content-Length:0 with 200 OK is therefore an edge case I'd like to resolve cleanly for this spec as not being permitted and fof 204 to be used instead.

Reply:
Thanks for pointing to the updated spec. I agree that it is an edge case. As this is a new standard, there's no problem with setting some firmer ground rules.

Reply:
In his [webinar about callback support in OpenAPI 3.0](https://t.co/1XUELjz5G5), @fehguy suggests to use 204 response as one way to signal that receiving party is no longer interested in further updates, that it that callback to be unregistered from web hook (as alternative to explicit API for unregistering callbacks/subscriptions).

Reply:
@sslavic I don't agree that 204 telegraphs any such intent, ever. If you want to terminate a relationship you need to reach into the 3xx bucket.